rust/sbp/src/codec.rs

```diff
@@ -344,11 +344,40 @@ pub mod json {
                 other: input.other,
             };
 
-            let mut writer = dst.writer();
-            serde_json::to_writer(&mut writer, &output)?;
+            let mut ser = Serializer::with_formatter(dst.writer(), HaskellishFloatFormatter {});
+            output.serialize(&mut ser)?;
             dst.put_slice(b"\n");
 
             Ok(())
         }
     }
+
+    /// Provide Haskell style formatting (sort of).  See `haskellish_float`.
+    struct HaskellishFloatFormatter {}
+
+    impl Formatter for HaskellishFloatFormatter {
+        #[inline]
+        fn write_f32<W: ?Sized>(&mut self, writer: &mut W, value: f32) -> std::io::Result<()>
+        where
+            W: std::io::Write,
+        {
+            if value == 0.0 || value.abs() >= 0.1 && value.abs() <= 9_999_999.0 {
+                write!(writer, "{}", value)
+            } else {
+                write!(writer, "{:e}", value)
+            }
+        }
+
+        #[inline]
+        fn write_f64<W: ?Sized>(&mut self, writer: &mut W, value: f64) -> std::io::Result<()>
+        where
+            W: std::io::Write,
+        {
+            if value == 0.0 || value.abs() >= 0.1 && value.abs() <= 9_999_999.0 {
+                write!(writer, "{}", value)
+            } else {
+                write!(writer, "{:e}", value)
+            }
+        }
+    }
```

Comment:
Not sure if there is a way to avoid this duplication without defining a float trait (or pulling one in)

Reply:
You could define a macro